### ISSUE TITLE
Unconventional use of Keras call API

### ISSUE THREAD

**Issue by cgarciae** [tensorflow/models#8018]
In the `Attention` layer, `call` expects multiple arguments and even keyword arguments but by convention keras uses a single `inputs` argument that can be a list of tensors to accept multiple arguments:

https://github.com/tensorflow/models/blob/3bb3f185c14c6c2db8131cfbb438ed06b5633169/official/transformer/v2/attention_layer.py#L71

**Comment by saberkun** [comment]
Yes, this is not conventional and it is a new feature in tf.keras.
Note, we have a set of new modeling library using keras convention more closely: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/layers/attention.py

In TF 2.0, you can use multiple positional args for inputs in keras sub-class Model. The main reason for this is the optional cache dictionary for transformer decoding. Because the cache is not used in training, the computation topology of keras model are different when it is used in training vs prediction.
In TF 2.1+, you can use multiple positional args for inputs with functional API.

**Comment by cgarciae** [comment]
@saberkun thanks for sharing this information. You have any link where I can read about positional and keyword arguments in `call` in TF 2.1+ ? 

I think there is a special `mask` kwarg simiar to `training`, maybe you can that for `attention_mask`?

**Comment by cgarciae** [comment]
Also, now that Multi Head (Self) Attention is becoming more mainstream, are there plans to add this layer to `tf.keras`? Currently there is an `Attention` layer but it only computes the main equation for a single head.

**Comment by saberkun** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/base_layer.py#L632
Sorry, I am probably wrong. I know that's the direction to move but maybe the args/kwargs issue is not fully solved for functional style now. @omalleyt12 

Yes, we will talk to keras team to bring some of these layers to tf.keras.

**Comment by saberkun** [comment]
@cgarciae I am working on moving it the keras. The prototype is hosting here: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/layers/attention.py
I will get an RFC soon.

**Comment by saberkun** [comment]
Keras layer support multiple arg/kwargs inputs now.
